python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -0,0 +1,381 @@
+        distribution.
+
+        Parameters
+        ----------
+        prob:  Tensor
+            Tensor contains a set of probabilities for each possible
+            class label of some model.
+        ori_pred: int
+            Integer representing the original prediction.
+
+        Returns
+        -------
+        float
+            The function returns the sum of the three loss components,
+            which is a scalar tensor representing the total loss.
+        """
+        logit = prob[ori_pred]
+        # 1e-6 added to logit to avoid taking the logarithm of zero
+        logit += 1e-6
+        # computing the cross-entropy loss for a single prediction
+        pred_loss = -torch.log(logit)
+
+        # size
+        edge_mask = self.sparse_mask_values
+        if self.coff_budget <= 0:
+            size_loss = self.coff_budget * torch.sum(edge_mask)
+        else:
+            size_loss = self.coff_budget * F.relu(
+                torch.sum(edge_mask) - self.coff_budget
+            )
+
+        # entropy
+        scale = 0.99
+        edge_mask = self.edge_mask * (2 * scale - 1.0) + (1.0 - scale)
+        mask_ent = -edge_mask * torch.log(edge_mask) - (
+            1 - edge_mask
+        ) * torch.log(1 - edge_mask)
+        mask_ent_loss = self.coff_connect * torch.mean(mask_ent)
+
+        loss = pred_loss + size_loss + mask_ent_loss
+
+        return loss
+
+    def concrete_sample(self, log_alpha, beta=1.0, training=True):
+        r"""Sample from the instantiation of concrete distribution when training.
+
+        Parameters
+        ----------
+        log_alpha : Tensor
+            A tensor representing the log of the prior probability of activating the gate.
+        beta : float, optional
+            Controls the degree of randomness in the gate's output.
+        training : bool, optional
+            Indicates whether the gate is being used during training or evaluation.
+
+        Returns
+        -------
+        Tensor
+            If training is set to True, the output is a tensor of probabilities that
+            represent the probability of activating the gate for each input element.
+            If training is set to False, the output is also a tensor of probabilities,
+            but they are determined solely by the log_alpha values, without adding any
+            random noise.
+        """
+        if training:
+            bias = self.sample_bias
+            random_noise = torch.rand(log_alpha.size()).to(log_alpha.device)
+            random_noise = bias + (1 - 2 * bias) * random_noise
+            gate_inputs = torch.log(random_noise) - torch.log(
+                1.0 - random_noise
+            )
+            gate_inputs = (gate_inputs + log_alpha) / beta
+            gate_inputs = torch.sigmoid(gate_inputs)
+        else:
+            gate_inputs = torch.sigmoid(log_alpha)
+
+        return gate_inputs
+
+    def train_explanation_network(self, dataset, func_extract_feat):
```

Comment:
Personally I prefer having this function only handle one forward step, i.e.,

```
feat = func_extract_feat(g)
prob, edge_mask = self.explain_graph(g, feat, tmp=tmp, training=True)
self.edge_mask = edge_mask
loss_tmp = self.loss(prob.unsqueeze(dim=0), ori_pred_dict[idx])
```
This leaves more flexibility to users. They can choose whatever learning rate, optimizer, number of epochs, etc., outside of this module.

Reply:
The users choose the learning rate, number of epochs, etc, but not the optimizer, and also the user can define their own network to train and not use our predefined network. 

So, should the user pass in a model for learning the edge mask, also? 

"You can move the rest code to an example." -- Sorry, didn't understand this part, should I make a new module or just update the testing script?

Reply:
I'm thinking of something like below.

```
explainer = PGExplainer(...)
optimizer = ...

# Train the explanation network.
for epoch in range(num_epochs):
      for g in graph_dataloader (graph_dataset):
            loss = explainer.train_step(g)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

# Explain.
explainer.explain(some_graph)
```

This gives more flexibility to users and makes the module more light-weight.

Reply:
> "You can move the rest code to an example."

I meant the example section in the module doc.

Reply:
Understood, thank you. Will push the changes soon.

Reply:
@mufeili comments addressed, please take a look.